src/SubPool.sol

```diff
@@ -0,0 +1,157 @@
+    }
+
+    /// @notice withdraw arbitrary tokens
+    /// @dev Can only withdraw non cow and non collateral tokens while the pool is active or in exit delay.
+    ///      When the pool's exit delay has elapsed it can withdraw any token and ether balance.
+    function withdrawTokens(address[] calldata tokens) external auth {
+        uint256 exitTimestamp = factory.exitTimestamp(address(this));
+        bool exitElapsed = exitTimestamp != 0 && block.timestamp >= exitTimestamp;
+
+        if (exitElapsed) {
+            for (uint256 i = 0; i < tokens.length;) {
+                SafeTransferLib.safeTransfer(tokens[i], msg.sender, ERC20(tokens[i]).balanceOf(address(this)));
+                unchecked {
+                    ++i;
+                }
+            }
+            uint256 ethBalance = address(this).balance;
+            if (ethBalance > 0) {
+                SafeTransferLib.safeTransferETH(msg.sender, address(this).balance);
+            }
+        } else {
+            for (uint256 i = 0; i < tokens.length;) {
+                address token = tokens[i];
+                if (token == COW || token == collateralToken) revert SubPool__InvalidWithdraw();
+                SafeTransferLib.safeTransfer(token, msg.sender, ERC20(token).balanceOf(address(this)));
+                unchecked {
+                    ++i;
+                }
+            }
+        }
+    }
+
+    /// @notice Pull required number of tokens from the sender to push the pool
+    ///         above collateralization.
+    function heal() external payable {
+        uint256 amt = collateralDue;
+        uint256 cowAmt = cowDue;
+        uint256 ethAmt = ethDue;
+        payDues(amt, cowAmt, ethAmt);
+    }
+
+    /// @notice Pay partial or full dues.
+    function payDues(uint256 amt, uint256 cowAmt, uint256 ethAmt) public payable {
+        if (msg.value != ethAmt) revert SubPool__InsufficientETH();
+        if (amt > 0) {
+            SafeTransferLib.safeTransferFrom(collateralToken, msg.sender, address(this), amt);
+            uint256 due = collateralDue;
+            collateralDue = amt >= due ? 0 : due - amt;
+        }
+        if (cowAmt > 0) {
+            SafeTransferLib.safeTransferFrom(COW, msg.sender, address(this), cowAmt);
+            uint256 due = cowDue;
+            cowDue = cowAmt >= due ? 0 : due - cowAmt;
+        }
+        if (ethAmt > 0) {
+            uint256 due = ethDue;
+            ethDue = ethAmt >= due ? 0 : due - ethAmt;
+        }
+    }
+
+    /// @notice Bill some fines.
+    function bill(uint256 amt, uint256 cowAmt, uint256 ethAmt, address to) external onlyFactory {
+        // verify that the pool's exit delay has not elapsed if it was requested.
+        uint256 exitTimestamp = factory.exitTimestamp(address(this));
+        if (exitTimestamp != 0 && block.timestamp >= exitTimestamp) revert SubPool__CannotBillAfterExitDelay();
+
+        if (amt > 0) {
+            SafeTransferLib.safeTransfer(collateralToken, to, amt);
```

Comment:
What if this fails? i think we should still be able to bill more than is currently held.

Reply:
I think there should be a distinction between billing ("dear solver, you owe us, and this is the amount; you won't be able to solve until you pay") and fining ("dear solver, you screwed up, your money is mine now").

Reply:
If you allow unpaid dues, you will have to keep track of those, have more conditions on exits to check those are paid or retained before rest of withdrawals are paid out, etc. Do we want all that complexity?

Reply:
@fleupold @fedgiac @mfw78

Reply:
I don't think this makes any sense to be able to bill more than is held. There is already a _tonne_ of off-chain stuff, and billing more than what is held sounds like a pretty serious issue - so there will likely already be off-chain aspects happening at this time, so I don't see any point in enshrining within the contract this increased complexity.

Reply:
ok, let's just be able to take what's in there

Reply:
No change here